packages/meditrak-server/src/routes/importSurveyResponses/SurveyResponseUpdatePersistor.js

```diff
@@ -0,0 +1,219 @@
+  const batches = [];
+  for (let i = 0; i < deletes.length; i += RECORDS_PER_BULK_BATCH) {
+    batches.push(deletes.slice(i, i + RECORDS_PER_BULK_BATCH));
+  }
+  return batches;
+}
+
+/**
+ * This class will hold on to all of the creates, updates, and deletes that a survey response import
+ * process wants to make, and allow running them all in batches once the whole spreadsheet is parsed
+ *
+ * The idea is that we can separate out the actual database updates from the parsing, validation,
+ * and formatting elements of the process. Those earlier parts can then run and synchronously return
+ * to the user with any validation errors. The db updates can then run in batches in the background.
+ */
+export class SurveyResponseUpdatePersistor {
+  constructor(models) {
+    this.models = models;
+    // maintain a map of response ids to update details, where update details looks like
+    // { sheetName, surveyResponseId, type, newSurveyResponse, newDataTime, answersToUpsert, answersToDelete }
+    // n.b. all but sheetName, surveyResponseId, and type are optional
+    this.updatesByResponseId = {};
+  }
+
+  count() {
+    return Object.keys(this.updatesByResponseId).length;
+  }
+
+  setupColumnsForSheet(sheetName, surveyResponseIds) {
+    surveyResponseIds.forEach((surveyResponseId, columnIndex) => {
+      if (!surveyResponseId) return; // array contains some empty slots representing info columns
+      this.updatesByResponseId[surveyResponseId] = {
+        type: UPDATE,
+        sheetName,
+        columnIndex,
+        surveyResponseId,
+        newSurveyResponse: null, // only populated if a new survey response is to be created
+        newDataTime: null, // only populated if submission time is to be updated
+        answers: {
+          upserts: [], // populated for new or updated survey responses
+          deletes: [], // populated for existing survey responses if answers are to be deleted
+        },
+      };
+    });
+  }
+
+  createSurveyResponse(surveyResponseId, newSurveyResponse) {
+    this.updatesByResponseId[surveyResponseId].type = CREATE;
+    this.updatesByResponseId[surveyResponseId].newSurveyResponse = newSurveyResponse;
+  }
+
+  deleteSurveyResponse(surveyResponseId) {
+    this.updatesByResponseId[surveyResponseId].type = DELETE;
+  }
+
+  updateDataTime(surveyResponseId, newDataTime) {
+    this.updatesByResponseId[surveyResponseId].newDataTime = newDataTime;
+  }
+
+  upsertAnswer(surveyResponseId, details) {
+    this.updatesByResponseId[surveyResponseId].answers.upserts.push(details);
+  }
+
+  deleteAnswer(surveyResponseId, details) {
+    this.updatesByResponseId[surveyResponseId].answers.deletes.push(details);
+  }
+
+  // ---- process creates in bulk, batching to avoid running out of memory ----
+
+  async processCreates(creates) {
+    const failures = [];
+    for (const batchOfCreates of batchCreates(creates)) {
+      try {
+        await this.models.wrapInTransaction(async transactingModels => {
```

Comment:
Yeah I think it makes things a little tricky for the users as they have to work out which to retry and which have succeeded (they get that info in an email, but have to update their spreadsheet). The bright side is that we should hopefully never have errors within this transaction if we've got good enough validation, because there ideally wouldn't be database level errors 🤞